blocks/http/types.h

```diff
@@ -88,16 +88,97 @@ struct GET : HTTPRequestBase<GET> {
   explicit GET(const std::string& url) : HTTPRequestBase(url) {}
 };
 
-struct ChunkedGET {
+template <typename T>
+struct ChunkedBase {
   const std::string url;
-  std::function<void(const std::string&, const std::string&)> header_callback;
-  std::function<void(const std::string&)> chunk_callback;
-  std::function<void()> done_callback;
-  explicit ChunkedGET(const std::string& url,
-                      std::function<void(const std::string&, const std::string&)> header_callback,
-                      std::function<void(const std::string&)> chunk_callback,
-                      std::function<void()> done_callback = []() {})
-      : url(url), header_callback(header_callback), chunk_callback(chunk_callback), done_callback(done_callback) {}
+
+  const std::function<void(const std::string&, const std::string&)> header_callback;
+  const std::function<void(const std::string&)> chunk_callback;
+  const std::function<void()> done_callback;
+
+  std::vector<std::unique_ptr<current::strings::StatefulGroupByLines>> group_by_lines_;
+
+  std::function<void(const std::string&, const std::string&)> header_callback_impl;
+  std::function<void(const std::string&)> chunk_callback_impl;
+  std::function<void()> done_callback_impl;
+
+  explicit ChunkedBase(std::string url)
+    : url(std::move(url)),
+      header_callback([this](const std::string& k, const std::string& v) { header_callback_wrapper(k, v); }),
+      chunk_callback([this](const std::string& c) { chunk_callback_wrapper(c); }),
+      done_callback([this]() { done_callback_wrapper(); }) {}
+
+  explicit ChunkedBase(std::string url,
+                       std::function<void(const std::string&, const std::string&)> header_callback,
+                       std::function<void(const std::string&)> chunk_callback,
+                       std::function<void()> done_callback = []() {})
+      : ChunkedBase(std::move(url)) {
+    header_callback_impl = header_callback;
+    chunk_callback_impl = chunk_callback;
+    done_callback_impl = done_callback;
+  }
+
+  T& OnHeader(std::function<void(const std::string&, const std::string&)> header_callback) {
+    this->header_callback_impl = header_callback;
+    return static_cast<T&>(*this);
+  }
+  T& OnChunk(std::function<void(const std::string&)> chunk_callback) {
+    this->chunk_callback_impl = chunk_callback;
+    return static_cast<T&>(*this);
+  }
+  T& OnDone(std::function<void()> done_callback) {
+    this->done_callback_impl = done_callback;
+    group_by_lines_.clear();
+    return static_cast<T&>(*this);
+  }
+
+  // TODO(dkorolev): Move this to `Chunk`-s if performance becomes the bottleneck.
+  T& OnLine(std::function<void(const char*)> line_callback) {
```

Comment:
Hmm, not sure why the API for lines is `const char*`, while other functions are `const std::string&`

Reply:
Right, let's keep them `const std::string&`-s to keep things simple.